Add Llama Flax Implementation

vvvm23 opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

Fixes #26809. This is a work-in-progress port of Llama to Flax, leaving it as a draft PR for now.

The implementation is based heavily off the GPT-Neo and GPT-J Flax implementations.

Currently, the submodules are ready, I just need to assemble into a full model, check weight loading, add tests, and update the documentation.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. [mentioned in this issue comment](https://github.com/huggingface/transformers/issues/22647#issuecomment-1579154174)
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @sgugger

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam

Documentation: @sgugger, @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: @sgugger
- TensorFlow: @Rocketknight1

 -->

vvvm23 committed:
Copies `modeling_flax_gpt_neo.py` to start

vvvm23 committed:
MLP Block. WIP Attention and Block

vvvm23 committed:
Adds Flax implementation of `LlamaMLP`
Validated with in-file test.
Some slight numeric differences, but assuming it isn't an issue

vvvm23 committed:
Adds `FlaxLlamaRMSNorm` layer
`flax.linen` includes `RMSNorm` layer but not necessarily in all
versions. Hence, we add in-file.

vvvm23 committed:
Adds FlaxLlamaAttention
Copied from GPT-J as it has efficient caching implementation as well as
rotary embeddings.
Notice numerically different, but not by a huge amount. Needs
investigating

vvvm23 committed:
Adds `FlaxLlamaDecoderLayer`
numerically inaccurate, debugging..

vvvm23 committed:
debugging rotary mismatch
gptj uses interleaved whilst llama uses contiguous
i think they match now but still final result is wrong.
maybe drop back to just debugging attention layer?

vvvm23 committed:
fixes bug with decoder layer
still somewhat numerically inaccurate, but close enough for now

vvvm23 committed:
adds markers for what to implement next
the structure here diverges a lot from the PT version.
not a big fan of it, but just get something working for now

sanchit-gandhi commented:
Very cool @vvvm23! Scanned through the PR and it looks very nice already - happy to do a full review when it's close to completion. Just drop me a line and I'll have a look! 🚀 Likewise if you have any questions or queries, I'm on hand to help :)

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_24587). All of your documentation changes will be reflected on that endpoint.

gianlucadetommaso commented:
Hi @vvvm23 and @sanchit-gandhi, do you guys have a timeline for this effort? Asking because I would love to import FlaxLlama from Hugging Face, but if it is going to take a while, I will probably build my own pipeline to import the model.

Not sure if this helps at all, but [here](https://github.com/young-geng/EasyLM/blob/main/EasyLM/models/llama/llama_model.py) you find an implementation of Llama in Flax (plus some other library-specific methods that you probably won't need).

vvvm23 commented:
Hi @gianlucadetommaso, I haven't had the time to work on this since this draft PR went live, but I am blocking time out this weekend to continue.

sanchit-gandhi commented:
Cool to see community interest around running Flax Llama! Feel free to ping me here when you need a review @vvvm23!

vvvm23 committed:
implements `FlaxLlamaBlockCollection`]
tolerance must be higher than expected, kinda disconcerting

vvvm23 commented:
Thanks @sanchit-gandhi I found a little time to continue today.

One issue I am noticing is that the tolerance when comparing the ground truth PyTorch implementation (in `modeling_llama.py`) and my own implementation, is a lot higher than I'd like. For three hidden layers in the decoder stack, I have to raise it to `atol=1e-2, rtol=1e-2`, with one hidden layer being at `atol=1e-3, rtol=1e-3` in order to pass. You can see the scratch test I am using at the bottom of `modeling_flax_llama.py`

I think some numerical differences are expected, but not sure to what degree. I am also testing with `float32` so that made me even more suspicious. Would you expected the results to be identical? This is my first time porting a PyTorch model to Flax. Thanks~

vvvm23 committed:
Adds `FlaxLlamaModule`
equivalent PyTorch model is `LlamaModel`
yay! a language model🤗

vvvm23 committed:
adds `FlaxLlamaForCausalLMModule`
equivalent to `LlamaForCausalLM`
still missing returning dict or tuple, will add later

vvvm23 commented:
Update: I now have a full model working. I haven't checked if the pretrained weight loading wrappers (provided by the Flax GPTNeo implementation) work yet, but once they are it will be ready for review. I'll simultaneously clean it up and add some missing features whilst it is being reviewed.

vvvm23 committed:
start porting pretrained wrappers
realised it probably needs return dict as a prereq

sanchit-gandhi commented:
Hey! Thanks for the progress update here @vvvm23 and great questions regarding numerical equivalence between models.

Generally, for any model less than 1B params we should be able to get equivalence to within 1e-5 between Flax and PyTorch. It's quite likely that you won't get this equivalence running the matmuls in bfloat16 on TPU. But you should be able to running the matmuls in float32, see https://github.com/huggingface/transformers/issues/15754 and https://github.com/google/jax/issues/10413#issue-1212211265 for details

Here's a script that I used previously for checking PT / Flax equivalence for BLOOM: https://github.com/sanchit-gandhi/codesnippets/blob/main/check_flax_bloom_jit_small_testing.ipynb You can ignore the bits about JIT'ing the forward pass for the time being. You can also uncomment the check to run it on CPU to force the highest precision, or use the decorator as provided

vvvm23 committed:
cleanup, quality, style

vvvm23 committed:
readds `return_dict` and model output named tuples

vvvm23 committed:
(tentatively) pretrained wrappers work 🔥

vvvm23 committed:
fixes numerical mismatch in `FlaxLlamaRMSNorm`
seems `jax.lax.rsqrt` does not match `torch.sqrt`.
manually computing `1 / jax.numpy.sqrt` results in matching values.

vvvm23 commented:
I've begun my hunt for numerical bugs 🐛

The first I squashed was rather strange. It seems `torch.rsqrt` and `jax.lax.rsqrt` do not match. This is used in the RMSNorm layers. Simple test to reproduce:
```
In [19]: a = np.asarray(a, dtype=np.float32)

In [20]: a
Out[20]:
array([1.16661310, 1.46686172, 0.13794081, 1.22346771, 1.17509305],
      dtype=float32)
In [21]: torch.rsqrt(torch.from_numpy(a))
Out[21]: tensor([0.92584139, 0.82566792, 2.69248700, 0.90407354, 0.92249471])

In [22]: jax.lax.rsqrt(a)
Out[22]: Array([0.92584133, 0.82566792, 2.69248700, 0.90407354, 0.92249471],      dtype=float32)

In [23]: 1 / torch.sqrt(torch.from_numpy(a))
Out[23]: tensor([0.92584139, 0.82566792, 2.69248700, 0.90407354, 0.92249471])

In [24]: 1 / jax.numpy.sqrt(a)
Out[24]: Array([0.92584139, 0.82566792, 2.69248700, 0.90407354, 0.92249471],      dtype=float32)
```
So the fix there was just to replace the `jax.lax.rsqrt` calls with `1 / jax.numpy.sqrt(...)`

Models still mismatches so I'll keep digging.

vvvm23 committed:
[WIP] debugging numerics

vvvm23 committed:
numerical match
I think issue was accidental change of backend. forcing CPU fixes test.
We expect some mismatch on GPU.

vvvm23 commented:
@sanchit-gandhi The model now numerically matches in fp32 on CPU. The issue was my backend has changed from CPU to GPU since fixing the `rsqrt` issue. I don't think we can expect a perfect match on GPU as the two models use fundamentally different backends. If there is anything you know of that could help remedy this, let me know.

What are the next steps to take? I am guessing some model tests, as well as trying it out on a real model checkpoint rather than random weights. However, my dev machine goes OOM when attempting to load the checkpoint on CPU.

sanchit-gandhi commented:
Hey @vvvm23! Excellent work on pinpointing the difference between torch and jax.lax `rsqrt` and glad to hear we're within numerical precision using fp32 on CPU - we can be pretty confident we have an accurate Flax implantation based on these results. For GPU, there will be differences between PyTorch and JAX. This is expected since JAX fundamentally works differently to PyTorch with how it computes the matmuls, and is OK since the JAX model will typically generate predictions that are 'as good' as the PyTorch one.

Adding some tests and updating the docs would be the most sensible next steps! Again, you can refer to the Flax GPT Neo model to see the relevant tests to add: https://github.com/huggingface/transformers/blob/main/tests/models/gpt_neo/test_modeling_flax_gpt_neo.py

> However, my dev machine goes OOM when attempting to load the checkpoint on CPU.

That's interesting - are we loading the weights on GPU by accident? There shouldn't be any GPU OOM if running on CPU. We might see our RAM get full if loading extremely large weights, but the GPU memory shouldn't be affected. What model size are you loading? We can try the smallest 7b checkpoint: https://huggingface.co/meta-llama/Llama-2-7b

vvvm23 commented:
Awesome thanks, tests and docs it is! I am currently on leave so won't be progressing on this until the 31st.

> That's interesting - are we loading the weights on GPU by accident? 
Actually, in the end no. By OOM on my dev machine, I meant out of CPU memory. Switching to a GPU backend meant I could load the model without running out of memory. So, nothing to worry about 😅

sanchit-gandhi commented:
Awesome - thanks for the update @vvvm23. Looking forward to doing a full review of the PR on your return!

vvvm23 committed:
adds in model and integration tests for Flax Llama
summary of failing:
- mul invalid combination of dimensions
- one numerical mismatch
- bf16 conversion (maybe my local backend issue)
- params are not FrozenDict

vvvm23 committed:
adds missing TYPE_CHECKING import and `make fixup`

vvvm23 committed:
adds back missing docstrings
needs review on quality of docstrings, not sure what is required.
Furthermore, need to check if `CHECKPOINT_FOR_DOC` is valid. See TODO

vvvm23 committed:
commenting out equivalence test as can just use common

sanchit-gandhi commented:
How's it looking @vvvm23? Let me know if I can help in anyway! Otherwise feel free to ping me here as soon as we're ready for a review, very excited to add this Flax model for the community!

vvvm23 commented:
Hi, currently been pretty split responsibility wise (moving house and job !!) so have only made a small bit of progress.

Most of the tests pass, however, there seems to be some matmul shape mismatch in the `generate_*` tests. Guessing I didn't implement the KV cache correctly, so I'll need to look at that. I also added back some missing docstrings.

I'll have some time to work on this Thursday, Friday (10th and 11th) but then probably nothing for another week :exploding_head: If you are in a rush and fancy trying to get the remaining tests to pass, please try! Sorry for the slowness on my part also!

vvvm23 committed:
debugging

vvvm23 committed:
Fixes bug where mask and pos_ids were swapped in pretrained models
This results in all tests passing now 🔥

vvvm23 commented:
The final tests ended up being easy to fix: I had simply forgotten to swap the attention mask and position ids in the pretrained model wrapper.

@sanchit-gandhi I haven't retested the slow tests locally (as my laptop is slow) but later today I can run them, then tidy the code a bit. If all goes well, should be good for review later today or early tomorrow 👍

vvvm23 committed:
cleanup of modeling file

vvvm23 committed:
cleanup of test file

vvvm23 commented:
@sanchit-gandhi all tests pass locally :tada: And I've also ran the model using the `generate` API to see if the outputs make sense:
```
In [23]: inputs = tokenizer('Aloha, World!', return_tensors='np')

In [24]: tokenizer.decode(model.generate(**inputs, generation_config=model.generation_config, max_length=100).sequences[0])
Out[24]: '<s> Aloha, World!\nI’m back from my trip to Hawaii and I’m feeling great! I’m still trying to get back into the swing of things, but I’m getting there. I’m going to be posting a lot of pictures from my trip, so stay tuned!\nI’m also going to be posting a lot of pictures from my trip to Hawaii, so stay tuned!\nI’m also going to be posting a lot of pictures'
```
Seems good to me!

---

I think this is ready for review. I would like to draw your attention to a few points I was unsure about:

Firstly, the model currently throws a warning when loading pretrained checkpoints:
```
Some weights of the model checkpoint at openlm-research/open_llama_3b_v2 were not used when initializing FlaxLlamaForCausalLM: 
{('model', 'layers', '9', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '1', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '24', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '11', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '7', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '23', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '13', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '5', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '6', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '20', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '21', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '16', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '10', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '4', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '0', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '25', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '12', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '3', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '19', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '14', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '18', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '22', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '8', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '15', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '17', 'self_attn', 'rotary_emb', 'inv_freq'), ('model', 'layers', '2', 'self_attn', 'rotary_emb', 'inv_freq')}
```
This has no effect on the outputs, just simply the Flax version of the model does not store the `inv_freq` tensor for rotary embeddings within the state dictionary, so these just get discarded. Is there a way to suppress this warning so not to scare any users?

Secondly, please double check the licensing. I just copied this from the PyTorch version of Llama and updated the year.

Third, I use the checkpoint `openlm-research/open_llama_3b_v2` as it was the smallest, fully open Llama checkpoint I could find. The 'official' Llama checkpoints have gated access, so I am unsure if they are appropriate for testing / documentation purposes. This also means I haven't been able to test the model with the official Llama checkpoint as I still haven't managed to get permission from Meta :cry: 

Fourth, as we discussed a lot of the code is copied from the Flax implementation of GPT-Neo. There may be some leftover parts from there that we don't need in Llama, and I may have missed some best practices for Llama as GPT-Neo is (relatively) old now. In particular, see the following code block in `FlaxLlamaPreTrainedModel.__call__`:

```
        # TODO: can this handle input tensors being passed as kwargs? I copied GPT-Neo directly here
        outputs = self.module.apply(
            inputs,
            jnp.array(input_ids, dtype="i4"),
            jnp.array(attention_mask, dtype="i4"),
            jnp.array(position_ids, dtype="i4"),
            not train,
            False,
            output_attentions,
            output_hidden_states,
            return_dict,
            rngs=rngs,
            mutable=mutable,
        )
```

Finally, the tests pass but please check if they have sufficient coverage :hugs: 

---

Generally speaking, I have a lot of experience writing model code but no experience making large contributions to the Huggingface ecosystem, so there is almost certainly a lot wrong! Apologies in advance and I will do my best to help you bring this model to the finish line :muscle: Thanks for your work so far!

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+        query = self.q_proj(hidden_states)
+        key = self.k_proj(hidden_states)
+        value = self.v_proj(hidden_states)
+        query = self._split_heads(query)
```
(nit) would add an extra line to break-up the projections and split heads for readability

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+        self.o_proj = dense()
+
+        self.causal_mask = make_causal_mask(jnp.ones((1, config.max_position_embeddings), dtype="bool"), dtype="bool")
+        self.embed_positions = create_sinusoidal_positions(config.max_position_embeddings, self.head_dim)
```
This style of embedding positions where we don't register them to the state dict works for GPT-J, since both the PyTorch and Flax models compute the embedding positions on the fly

However, for LLaMA, the PyTorch model registers the `inv_freq` to the state dict, and caches the sin/cos positional embeddings on the first forward pass

Now in reality this is fine, since these are not trained model weights, but rather stored constants that can be computed in the `__init__` / `setup` as required. However, we should bear in mind that we'll get weight mis-matches going from PT -> Flax, and **also** from Flax -> PT (i.e. if we load Flax weights in a PyTorch model)

Ideally, we should add these constants to the weight dictionary of the Flax model as well to silence these weight mis-matches -> do you think this is something you could achieve with a small refactor to the embeddings? It's not a big deal if we don't manage, but this would be the ideal situation

vvvm23 replied:
I think this is doable. We could just make it a variable in the flax model during setup and initialise it in the same way as we would on the fly.

vvvm23 replied:
@sanchit-gandhi I started looking into this, however the shape of the sinusoidal embeddings is different between the GPT-Neo Flax and Llama PyTorch implementations. Hence, once it is added to the parameter dictionary, it will throw an error during testing as the shapes are different.

We could refactor the Flax version to change the shape, or suppress the warning when there are missing weights for this specific parameter. The latter should be okay as there is no functional difference as far as I can tell. What do you think?

vvvm23 replied:
@sanchit-gandhi or, third option, like you said:
> It's not a big deal if we don't manage, but this would be the ideal situation

🤔

sanchit-gandhi replied:
Hey @vvvm23! Did you push code for registering the sinusoidal positional embeddings in the state dict? It would be cool to take a look if you have something to hand, but no worries if not! My preference would be for refactoring the Flax code to match the shape of the PyTorch weights, so that we can load them directly. But if this completely breaks the Flax weights / forward pass with some heavy re-shapes then maybe we can skip. What are the shapes currently in Flax and PyTorch?

vvvm23 replied:
Haven't yet, was waiting confirmation either way 😅 I can give it a shot tomorrow. I think the main difference is just the order of dimensions, just might affects masking and caching behaviour too but hopefully not in a tricky way. Will update you tomorrow.

vvvm23 replied:
@sanchit-gandhi the solution to this ended up being quite simple. I had misudnerstand what `inv_freq` did.

In the PyTorch code:
- `inv_freq` is a tensor of size `head_dim // 2` containing the reciprocal of increasing frequencies.
- This tensor is cached as part of the state dictionary.
- This tensor can be used to derive `cos_cached` and `sin_cached` which are also registered as buffers but with `persistent=False` so they do not become part of the saved state dictionary.

In my Flax code:
- `inv_freq` is calculated and immediately used to derive `cos_cached` and `sin_cached`.
- `inv_freq` is immediately discarded and overridden with the same name to the concatenation of `cos_cached` and `sin_cached`

So, I had mistakedly given `inv_freq` the wrong variable name – it is more like `sincos`.

So to fix, I just create the tensor `inv_freq` as in the PyTorch code. This is unused in the Flax, but will be used when loading back to a PyTorch model. This stops the warning.

Sorry for my mistake 😓

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+        else:
+            mutable = False
+
+        # TODO: can this handle input tensors being passed as kwargs? I copied GPT-Neo directly here
```
Can the **whole** of this class be copied from GPT-Neo? We can try prepending this before we define the class:
```
# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->Llama
```

vvvm23 replied:
I think it can, but I will verify this.

vvvm23 replied:
They are functionally identical, however seem to be formatted differently (seems GPT Neo had different `make style` settings when it was created). If it is formatted differently, is it okay to prepend that string?

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+
+
+class FlaxLlamaRMSNorm(nn.Module):
+    eps: float = 1e-6
```
I would rename to `epsilon` for consistency with Flax LayerNorm layers

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+
+    def setup(self):
+        hidden_size = self.config.hidden_size
+        inner_dim = self.config.intermediate_size if self.config.intermediate_size is not None else 4 * hidden_size
```
Let's just move this line to the `setup` of `FlaxLlamaMLP` and pass the config as argument

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+        self.input_layernorm = FlaxLlamaRMSNorm(eps=self.config.rms_norm_eps)
+        self.self_attn = FlaxLlamaAttention(self.config, dtype=self.dtype)
+        self.post_attention_layernorm = FlaxLlamaRMSNorm(eps=self.config.rms_norm_eps)
+        self.mlp = FlaxLlamaMLP(self.config, inner_dim, dtype=self.dtype)
```
```suggestion
        self.mlp = FlaxLlamaMLP(self.config, dtype=self.dtype)
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+        return outputs
+
+
+class FlaxLlamaBlockCollection(nn.Module):
```
Layer collection perhaps a better name?
```suggestion
class FlaxLlamaLayerCollection(nn.Module):
```

sanchit-gandhi replied:
In PyTorch, we have `nn.ModuleList` which we can use to combine N decoder layers together into one Torch `nn.Module`:
https://github.com/huggingface/transformers/blob/fd56f7f0813d412c3e0848cbd6f94a23de2c07b7/src/transformers/models/llama/modeling_llama.py#L566

There isn't an equivalent in Flax, so we have to define this dummy "collection" layer to string together N decoder layers into one block:
https://github.com/huggingface/transformers/blob/deaecdaf53cf4047b0c56e943a36d6a0e7aad38a/src/transformers/models/llama/modeling_flax_llama.py#L512-L515

It's fine to use "collection" since we use it consistently throughout the `transformers` repo, I just think the "block" bit could maybe be reworded to "layer"

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+    def setup(self):
+        self.hidden_size = self.config.hidden_size
+        embedding_init = jax.nn.initializers.normal(stddev=self.config.initializer_range)
+        self.embed_tokens = nn.Embed(
```
Should be fine to let the embedding layers use the dtype of the model (see for reference T5x: https://github.com/google-research/t5x/blob/ca3d2e43c8db2e6769073ffa98b7689443e3b2b8/t5x/examples/t5/network.py#L287)

sanchit-gandhi replied:
In fact this is required for bf16 forward pass

vvvm23 replied:
Good catch, this wasn't intentional 😅

vvvm23 replied:
I added the `dtype` parameter to the `nn.Embed` init call. However, the example you linked doesn't seem to match `flax.linen.Embed` - seems a custom implementation.

Should `param_dtype` also be equal to `self.dtype`?

sanchit-gandhi replied:
Good point! The T5x implementation builds on top of the generic Flax linen implementation. It's largely written and maintained by the same authors, but has some extra features on top for large model training & inference. We tend to use it as a reference for 'best practices' in Flax -> in this case, we can pass the `dtype` argument to the `nn.Embed` layer, but are fine to omit the `param_dtype` since this is handled outside the model when we explicitly upcast/downcast the params

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,202 @@
+    def __init__(
+        self,
+        parent,
+        batch_size=14,
```
Let's use bs=2 to keep the tests fast!

vvvm23 replied:
Nice, I would check the tests for GPT Neo and Llama and the batch sizes there too 👀

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,202 @@
+        use_input_mask=True,
+        use_token_type_ids=False,
+        use_labels=True,
+        vocab_size=99,
```
```suggestion
        vocab_size=99,
        hidden_size=16,
        num_hidden_layers=2,
        num_attention_heads=2,  
```

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,202 @@
+        num_hidden_layers=4,
+        num_attention_heads=4,
+        attention_types=[[["global", "local"], 2]],
+        intermediate_size=37,
```
Setting to hidden_size * 4. A factor of 128 should work better for testing with JAX if we ever run it on TPU
```suggestion
        intermediate_size=64,
```

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,202 @@
+
+@require_flax
+class FlaxLlamaModelTest(FlaxModelTesterMixin, FlaxGenerationTesterMixin, unittest.TestCase):
+    all_model_classes = (FlaxLlamaModel,) if is_flax_available() else ()
```
Let's make sure to test the ForCausalLM variant on the base tests
```suggestion
    all_model_classes = (FlaxLlamaModel, FlaxLlamaForCausalLM, ) if is_flax_available() else ()
```

vvvm23 replied:
Good catch!

vvvm23 replied:
Wait actually, isn't this class already tested in the line below it:
```python
    all_generative_model_classes = (FlaxLlamaForCausalLM,) if is_flax_available() else ()
```

My understanding was that the "model" classes were the ones without specific LM heads.

sanchit-gandhi replied:
Adding a model to `all_generative_model_classes` will test it for the generation tests, but not the base model tests. Likewise, adding a model to `all_model_classes` will test it for the base model tests, but not the generation tests.

=> to test a model with the base and generation tests, we need to add it to both lists

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,202 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
The tests look like a great start! Just to ensure we have equivalence with PyTorch, could we add two slow integration tests?
1. In the first test, can we run a forward pass and check that the logits are to within a small tolerance of the (correct) PyTorch ones? Similar to the following test (but using sensible inputs for the LLaMA model): https://github.com/huggingface/transformers/blob/eec5841e9f440c795fb9292d009675d97a14f983/tests/models/whisper/test_modeling_flax_whisper.py#L411
2. In the second test, could we generate a sequence, and verify that it matches what we get with generating using the PyTorch model? Similar to the test: https://github.com/huggingface/transformers/blob/eec5841e9f440c795fb9292d009675d97a14f983/tests/models/bloom/test_modeling_flax_bloom.py#L236

vvvm23 replied:
@sanchit-gandhi regarding these tests, should they be using real data (prompt written by me, tokenized by the correct pretrained tokenizer, then passed to the model) or is it sufficient to use dummy data?

vvvm23 replied:
Looking at other tests, it seems to not matter :) I will use some arbitrary tokens for testing.

sanchit-gandhi reviewed (commented):
Looking very nice already @vvvm23 - there's a lot that's right here! Great to hear that the tests are passing and that the generations look accurate. It should be quite fast to see this one to the finish line: just a small refactoring suggestion for the RMS norm layer, and two slow integration tests to check we have numerical correctness! Let's go!

vvvm23 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,701 @@
+    @nn.compact
+    def __call__(self, hidden_states):
+        input_dtype = hidden_states.dtype
+        variance = jnp.asarray(hidden_states, dtype=jnp.float32)
```
It is a valid suggestion, I simply copied this from GPT Neo and assumed it was like this for numeric stability (force an upcast for this op). You don't think there will be any issues with that here? Maybe not with bf16 but there could be with f16

sanchit-gandhi replied:
Took a deeper look into this - you're indeed correct that the layer norm statistics should be kept in at least float32 precision: https://github.com/google/flax/blob/9e82cdb0e91d5724ec03f7cc3eceb7494e592c01/flax/linen/normalization.py#L71-L72

What the Flax linen layer norm calculations do is compute the variance / mean stats in default precision (fp32), but then downcast the result down to whatever the dtype is of the model.

I think we can follow this pattern here - so we can specify a dtype of float32 for the variance / mean stats, and then cast the final result to the `dtype` attribute

vvvm23 committed:
Resolving simpler review comments

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,703 @@
+        self.epsilon = self.config.rms_norm_eps
+        self.weight = self.param("weight", lambda _, shape: jnp.ones(shape), self.config.hidden_size)
+
+    @nn.compact
```
We can remove this `nn.compact` entirely now that we have a set-up

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,703 @@
+    @nn.compact
+    def __call__(self, hidden_states):
+        input_dtype = hidden_states.dtype
+        variance = jnp.asarray(hidden_states, dtype=self.dtype)
```
What you had here with the dtype of float32 was correct before here! Apologies!

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,703 @@
+        # hidden_states = hidden_states * jax.lax.rsqrt(variance + self.epsilon)
+        hidden_states = hidden_states / jnp.sqrt(variance + self.epsilon)
+
+        return jnp.asarray(self.weight * hidden_states, dtype=input_dtype)
```
Perfect - casting back to the input dtype is exactly the behaviour we want!

vvvm23 commented:
Thanks for your additional comments, I have some time to work on the more involved points today 🤗

vvvm23 committed:
addresses more minor review comments

vvvm23 committed:
fixing introduced pytest errors from review

vvvm23 committed:
wip additional slow tests

vvvm23 committed:
wip tests
need to grab a GPU machine to get real logits for comparison
otherwise, slow tests should be okay

vvvm23 committed:
`make quality`, `make style`

vvvm23 committed:
`make fix-copies`

vvvm23 committed:
fix mangled function following `make fix-copies`

vvvm23 commented:
@sanchit-gandhi I think everything except the missing weight issue is resolved now (see my comment).

Trying to resolve some remaining CI issues, I noticed that the line `# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->Llama` will change the line `@add_start_docstrings_to_model_forward(GPT_NEO_INPUTS_DOCSTRING)` , and overwrite `LLAMA_INPUTS_DOCSTRNG`. Any idea how to stop this happening? Otherwise the CI won't pass :thinking:

vvvm23 committed:
adds slow integration tests
- checking logits
- checking hidden states
- checking generation outputs

vvvm23 committed:
adds missing type checking imports

sanchit-gandhi commented:
That's correct behaviour @vvvm23! What we need to do is create the variable `LLAMA_INPUTS_DOCSTRNG` that contains the necessary docstring info for LLAMA (more or less copied one-for-one from GPT Neo, but adapted for any different inputs)

vvvm23 commented:
Yeah, it is correct behaviour - what I meant though is that I did have a `LLAMA_INPUTS_DOCSTRING` in a previous commit, but running `make fix-copies` overwrote this docstring with the GPT-Neo version (as you suggested we add that at a class level). I guess my question was, how can we copy everything else in the class but somehow exclude the docstring line?

I get that we need the docstring itself, just currently the CI won't pass with both that docstring and the `# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->Llama` line. Does the issue make sense?

vvvm23 committed:
fixes missing parameter checkpoint warning

vvvm23 committed:
more finegrained 'Copied from' tags
avoids issue of overwriting `LLAMA_INPUTS_DOCSTRING`

vvvm23 commented:
@sanchit-gandhi fixed the CI issue (I think) by just adding more `Copied from ...` comments and deleting the class level comment. I also fixed the merge conflict. We should be good to go once CI passes I think 🙂

vvvm23 committed:
swaps import guards
??? how did these get swapped initially?

vvvm23 committed:
removing `inv_freq` again as pytorch version has now removed

vvvm23 committed:
attempting to get CI to pass

vvvm23 committed:
adds doc entries for llama flax models

vvvm23 committed:
fixes typo in __init__.py imports

vvvm23 committed:
adds back special equivalence tests
these come from the gpt neo flax tests. there is special behaviour for these models that needs to override the common version

vvvm23 commented:
@sanchit-gandhi the CI still fails, this is for two reasons. Could you assist me with resolving this?

1. The documentation test fails as it tries to load the checkpoint `_CHECKPOINT_FOR_DOC` however it needs `from_pt=True` to be set.
2. Flax Llama is based off the GPT Neo implementation. GPT Neo uses special tests to test equivalence between the flax and pytorch implementations. This overrides the common `test_equivalence_pt_to_flax` test. I copy these special tests (to make my flax tests pass). However, changing the tests for the flax version will cause the pytorch version to fail as it is using the flax version incorrectly.

All the tests pass and the model is pretty much ready. Just not sure how to get through these last two blockers. Help would be much appreciated!

vvvm23 committed:
overrides tests with dummy to see if CI passes
need to fill in these tests later

vvvm23 committed:
adds my contribution to docs

vvvm23 committed:
`make style; make quality`

sanchit-gandhi commented:
Awesome - thanks for the progress updates here @vvvm23. Agree with you that this is nearly in a position for a final review!

To fix your problems:
1. Could you open a pull request on the HF Hub to add the Flax model weights to the checkpoint? See this page for opening a PR: https://huggingface.co/openlm-research/open_llama_3b_v2/discussions (just click 'new pull request' and open a PR). You can ping me here once you've done this and I'll merge the Hub PR! Once this is done, we won't need the `from_pt` argument
2. Does the Flax-PyTorch test `test_equivalence_flax_to_pt` fail if we don't override it? I think it should be possible to use the model without this hack?

vvvm23 committed:
replaces random masking with fixed to work with flax version

vvvm23 commented:
> Could you open a pull request on the HF Hub to add the Flax model weights to the checkpoint?

PR is open 🙂 Git LFS is a great thing.

> You can ping me here once you've done this and I'll merge the Hub PR!

@sanchit-gandhi 

> Shall we just construct sensible attention masks in the model tester? I think it's the random attention mask that might be causing the problem

You are a genius! 😅 That was exactly the issue. I changed it to a fixed lower triangular tensor of ones (to test different lengths), padded with ones (to match size of the random one). This should give good enough coverage 👍

sanchit-gandhi commented:
That's awesome - happy to hear that fixed the issue! Let's wait until one of the OpenLM authors sees the open PR to add the flax weights: https://huggingface.co/openlm-research/open_llama_3b_v2/discussions/4#64f5b59fbce84cd8b1c80f38

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+        return (hidden_states,) + outputs[1:]
+
+
+class FlaxLlamaPreTrainedModel(FlaxPreTrainedModel):
```
```suggestion
# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->Llama, GPT_NEO->LLAMA, transformer->model
class FlaxLlamaPreTrainedModel(FlaxPreTrainedModel):
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+    base_model_prefix = "model"
+    module_class: nn.Module = None
+
+    # Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel.__init__ with GPTNeo->Llama
```
```suggestion
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+        module = self.module_class(config=config, dtype=dtype, **kwargs)
+        super().__init__(config, module, input_shape=input_shape, seed=seed, dtype=dtype, _do_init=_do_init)
+
+    # Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel.init_weights with GPTNeo->Llama
```
```suggestion
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+        else:
+            return random_params
+
+    # Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel.init_cache with GPTNeo->Llama
```
```suggestion
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+        return unfreeze(init_variables["cache"])
+
+    @add_start_docstrings_to_model_forward(LLAMA_INPUTS_DOCSTRING)
+    # Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel.__call__ with GPTNeo->Llama
```
```suggestion
```

sanchit-gandhi commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -287,6 +287,9 @@ def params(self, params: Union[Dict, FrozenDict]):
             params = unfreeze(params)
         param_keys = set(flatten_dict(params).keys())
         if len(self.required_params - param_keys) > 0:
+            import ipdb
```
Remember to remove this!

vvvm23 replied:
oh my 😱

sanchit-gandhi replied:
Ahaha no worries - it's easy to forget! Just in case you're interested: I used to use `ipdb` for debugging, but have recently migrated to PyCharm's debugger: https://www.jetbrains.com/help/pycharm/debugging-your-first-python-application.html

I find it far easier debugging with this vs `ipdb`, since you can define all of your breakpoints without changing the code, and inspect all the variables at once. There are similar debuggers available if you use VSCode, which you might find make you more efficient than using `ipdb`. Otherwise, if you find you're already fast with `ipdb`, it's a good option!

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,370 @@
+
+        input_mask = None
+        if self.use_input_mask:
+            input_mask = random_attention_mask([self.batch_size, self.seq_length])
```
Shall we use a non-random attention mask here for consistency with the PyTorch model test? Can use np trill

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,370 @@
+            )
+
+    @is_pt_flax_cross_test
+    def test_equivalence_pt_to_flax(self):
```
Probably don't need to override this test when we use a sensible attention mask

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,370 @@
+    # overwrite from common since `attention_mask` in combination
+    # with `causal_mask` behaves slighly differently
+    @is_pt_flax_cross_test
+    def test_equivalence_flax_to_pt(self):
```
Likewise

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,370 @@
+    def test_model_logits(self):
+        model_id = "openlm-research/open_llama_3b_v2"
+        model = FlaxLlamaForCausalLM.from_pretrained(model_id, from_pt=True)
+        test_batch = jnp.arange(32).reshape(4, 8) + 0x777
```
Ideally we explicitly write out this integer:
```python
        test_batch = jnp.arange(32).reshape(4, 8) + 1911
```

Even better would be if we used the inputs dict we've defined before:
```python
        _, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()

        logits = model(**inputs_dict).logits
```

But you'd need to update the expected values in this case

vvvm23 replied:
I'll leave with 1911 for now to avoid updating the logits – can come back to this if it is really needed.

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,370 @@
+
+        # fmt: off
+        EXPECTED_LOGITS = [-74.4243, -74.0680, -65.2507, -79.1658, -77.7460, -69.2379, -86.4588, -84.8933, -77.8456]
+        EXPECTED_MIN, EXPECTED_MAX, EXPECTED_MEAN = -96.9952, -18.4571, -65.0608
```
(nit) could you split these expected values over three lines please?

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,370 @@
+                for fx_output, pt_output in zip(fx_outputs, pt_outputs_loaded):
+                    self.assert_almost_equals(fx_output[:, -1], pt_output[:, -1].numpy(), 4e-2)
+
+    @slow
```
Could these slow tests have their own integration test class please? Like with Flax BLOOM: https://github.com/huggingface/transformers/blob/44d2c199f6c5fcc93b662d6d3aaa02f8e79c9c4b/tests/models/bloom/test_modeling_flax_bloom.py#L200

sanchit-gandhi commented on tests/models/llama/test_modeling_llama.py:
```diff
@@ -14,20 +14,22 @@
 from ...test_pipeline_mixin import PipelineTesterMixin
 
 
+if is_flax_available():
```
No need for this - let's remove

sanchit-gandhi commented on tests/models/llama/test_modeling_llama.py:
```diff
@@ -96,7 +98,8 @@ def prepare_config_and_inputs(self):
         input_mask = None
         if self.use_input_mask:
-            input_mask = random_attention_mask([self.batch_size, self.seq_length])
+            input_mask = torch.tril(torch.ones(7, 7))
```
What about just:
```python
            input_mask = torch.tril(torch.ones(self.batch_size, self.seq_length))
```

=> this would work no?

vvvm23 replied:
Yes, I forgot `tril` is defined for non-square tensors 😓

sanchit-gandhi reviewed (commented):
Thanks for iterating @vvvm23 🤗 Mainly just nits from me - looks close to completion! 

Does this implementation support LLaMA2? There are public checkpoints for LLaMA2 on the Hub that you can get access to by filling out this form: https://ai.meta.com/resources/models-and-libraries/llama-downloads/

And then download the HF weights here: https://huggingface.co/meta-llama/Llama-2-7b-hf

vvvm23 commented on docs/source/en/index.md:
```diff
@@ -313,7 +313,7 @@ Flax), PyTorch, and/or TensorFlow.
 |            CLIPSeg            |       ✅        |         ❌         |      ❌      |
 |            CodeGen            |       ✅        |         ❌         |      ❌      |
-|           CodeLlama           |       ✅        |         ❌         |      ❌      |
+|           CodeLlama           |       ✅        |         ❌         |      ✅      |
```
This auto populated when I ran the command to generate this table, so I can only assume so 😅

sanchit-gandhi replied:
Nice!

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 commented:
Hey @sanchit-gandhi, thanks for the second review~

I can try with Llama 2 weights, however last time I requested access I never got a response. I will try again when I get the chance.

Just a heads up, I have basically no time to work on this until the start of next week. I am imminently moving which is basically sucking up all my time currently 🙃 I am also without GPU for a bit. I ask for your continued patience 🙏 or if you have the time to perhaps tackle a couple of the smaller changes yourself? Thanks for your understanding~

vvvm23 committed:
updates `x`->`tensor` in `rotate_half`

vvvm23 committed:
addresses smaller review comments

vvvm23 committed:
Update docs/source/en/model_doc/llama.md

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
adds integration test class

vvvm23 committed:
adds `dtype` to rotary embedding to cast outputs

vvvm23 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+        return self.weight * jnp.asarray(hidden_states, dtype=input_dtype)
+
+
+class FlaxLlamaRotaryEmbedding(nn.Module):
```
good point, updated accordingly.

vvvm23 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,710 @@
+    return (tensor * cos_pos) + (rotate_half(tensor) * sin_pos)
+
+
+class FlaxLlamaRMSNorm(nn.Module):
```
I guess this would replace `input_dtype`? I've made a commit doing so.

vvvm23 committed:
adds type to flax llama rotary layer

vvvm23 committed:
`make style`

vvvm23 committed:
`make quality; make style`

vvvm23 commented:
Hey @sanchit-gandhi, thanks for bearing with me. I have addressed your comments. They were quite small but I didn't have the headspace to think about this earlier 😅 

Regarding Llama 2 weights, I got the Meta approval, so I can test with the Llama 2 weights. Unfortunately, I am still without a GPU workstation and will likely OOM on CPU. So if you want results there you will have to bear with me more – but we should be good to go as far as a "Llama 1" implementation goes.

vvvm23 committed:
`make fix-copies`

vvvm23 commented on docs/source/en/model_doc/llama.md:
```diff
@@ -48,7 +48,7 @@ come in several checkpoints they each contain a part of each weight of the model
 - The LLaMA tokenizer is a BPE model based on [sentencepiece](https://github.com/google/sentencepiece). One quirk of sentencepiece is that when decoding a sequence, if the first token is the start of the word (e.g. "Banana"), the tokenizer does not prepend the prefix space to the string.
 
-This model was contributed by [zphang](https://huggingface.co/zphang) with contributions from [BlackSamorez](https://huggingface.co/BlackSamorez). The code of the implementation in Hugging Face is based on GPT-NeoX [here](https://github.com/EleutherAI/gpt-neox). The original code of the authors can be found [here](https://github.com/facebookresearch/llama).
+This model was contributed by [zphang](https://huggingface.co/zphang) with contributions from [BlackSamorez](https://huggingface.co/BlackSamorez). The code of the implementation in Hugging Face is based on GPT-NeoX [here](https://github.com/EleutherAI/gpt-neox). The original code of the authors can be found [here](https://github.com/facebookresearch/llama). The Flax version of the implementation  ...
```
No worries :)

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
```
```suggestion
"""Flax LLaMA model."""
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+    rotate_half_tensor = jnp.concatenate(
+        (-tensor[..., tensor.shape[-1] // 2 :], tensor[..., : tensor.shape[-1] // 2]), axis=-1
+    )
+    """Rotates half the hidden dims of the input."""
```
```suggestion
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+        variance = jnp.power(variance, 2)
+        variance = variance.mean(-1, keepdims=True)
+        # use `jax.numpy.sqrt` as `jax.lax.rsqrt` does not match `torch.rsqrt`
+        # hidden_states = hidden_states * jax.lax.rsqrt(variance + self.epsilon)
```
```suggestion
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+    return (tensor * cos_pos) + (rotate_half(tensor) * sin_pos)
+
+
+class FlaxLlamaRMSNorm(nn.Module):
```
Nice!

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+        embed_dim = self.config.hidden_size
+        inner_dim = self.config.intermediate_size if self.config.intermediate_size is not None else 4 * embed_dim
+
+        jax.nn.initializers.normal(self.config.initializer_range)
```
Can you pass this as a `kernel_init` as in GPTJ? https://github.com/huggingface/transformers/blob/5e11d72d4d0939138fbabfebe9a69d2061519547/src/transformers/models/gptj/modeling_flax_gptj.py#L302

vvvm23 replied:
good catch

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+        self.up_proj = nn.Dense(inner_dim, use_bias=False, dtype=self.dtype)
+
+    def __call__(self, hidden_states):
+        hidden_states = self.up_proj(hidden_states) * self.act(self.gate_proj(hidden_states))
```
Feel free to split this over two lines - think it would help with readability

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+        return (hidden_states,) + outputs[1:]
+
+
+# Copied from transformers.models.gpt_neo.modeling_flax_gpt_neo.FlaxGPTNeoPreTrainedModel with GPTNeo->Llama, GPT_NEO->LLAMA, transformer->model
```
Nice!

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+        output_hidden_states: bool = False,
+        return_dict: bool = True,
+    ):
+        input_embeds = self.embed_tokens(input_ids.astype("i4"))
```
No need to cast the input ids here - this is done already in the `__call__` to `FlaxLlamaPreTrainedModel`, same as with GPTJ: https://github.com/huggingface/transformers/blob/5e11d72d4d0939138fbabfebe9a69d2061519547/src/transformers/models/gptj/modeling_flax_gptj.py#L482
```suggestion
        input_embeds = self.embed_tokens(input_ids)
```

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,717 @@
+    """,
+    LLAMA_START_DOCSTRING,
+)
+class FlaxLlamaForCausalLM(FlaxLlamaPreTrainedModel):
```
Can this be copied from GPTJ?
```suggestion
# Copied from transformers.models.gptj.modeling_flax_gptj.FlaxLlamaForCausalLM with GPTJ->Llama
class FlaxLlamaForCausalLM(FlaxLlamaPreTrainedModel):
```

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,264 @@
+    from transformers.models.llama.modeling_flax_llama import FlaxLlamaForCausalLM, FlaxLlamaModel
+
+
+if is_torch_available():
```
Let's just remove this if we're not importing torch

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,264 @@
+            hidden_dropout_prob=self.hidden_dropout_prob,
+            attention_probs_dropout_prob=self.attention_probs_dropout_prob,
+            max_position_embeddings=self.max_position_embeddings,
+            use_cache=False,
```
Why no cache? Let's re-enable since this will be used in most use cases

sanchit-gandhi commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,264 @@
+
+@slow
+@require_flax
+class FlaxLlamaIntegrationTest(unittest.TestCase):
```
Cool tests!

sanchit-gandhi reviewed (approved):
Looks super, nice work @vvvm23! Just some very minor style suggestions from me - drop me a message here when they're resolved (should be very fast now) and I'll request a final core maintainer review.

Re which checkpoint to use: let's stick with the open assistant one since it doesn't require any license

vvvm23 committed:
Apply suggestions from code review

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
applies suggestions from review

vvvm23 commented:
@sanchit-gandhi should be good to go now~ If there is anything else let me know, but I'll be at ICCV this week so maybe slow to respond :+1:

vvvm23 committed:
Update modeling_flax_llama.py

vvvm23 committed:
`make fix-copies`

sanchit-gandhi commented on tests/models/llama/test_modeling_llama.py:
```diff
@@ -97,7 +96,8 @@ def prepare_config_and_inputs(self):
         if self.use_input_mask:
-            input_mask = random_attention_mask([self.batch_size, self.seq_length])
+            input_mask = torch.tril(torch.ones(self.batch_size, self.seq_length))
+            # input_mask = torch.nn.functional.pad(input_mask, (0, 0, 0, self.batch_size - self.seq_length), value=1)
```
```suggestion
```

sanchit-gandhi commented on tests/models/llama/test_modeling_llama.py:
```diff
@@ -97,7 +96,8 @@ def prepare_config_and_inputs(self):
 
         input_mask = None
         if self.use_input_mask:
-            input_mask = random_attention_mask([self.batch_size, self.seq_length])
```
Note to reviewer: the Flax model requires a causal attention mask. It gives garbage outputs for inference with a non-causal attention masks (which is not a valid situation in real-life and thus one we don't have to worry about). We update the attention mask in the PyTorch model tester so that we can run our PT-Flax equivalence tests without overriding them.

sanchit-gandhi commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+        output_hidden_states: bool = False,
+        return_dict: bool = True,
+    ):
+        input_embeds = self.embed_tokens(input_ids)
```
Sorry my previous suggestion broke the tests 😅 Feel free to revert the changes @vvvm23 - this will fix the failing tests on the CI

```suggestion
       input_embeds = self.embed_tokens(input_ids.astype("i4"))
```

vvvm23 committed:
Update tests/models/llama/test_modeling_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
Update src/transformers/models/llama/modeling_flax_llama.py

Co-authored-by: Sanchit Gandhi <93869735+sanchit-gandhi@users.noreply.github.com>

vvvm23 committed:
fixes shape mismatch in FlaxLlamaMLP

sanchit-gandhi commented:
Failing tests are related to missing Flax weights for the checkpoint [openlm-research/open_llama_3b_v2](https://huggingface.co/openlm-research/open_llama_3b_v2).

Could you load the PyTorch weights in Flax, and then push them to the org in a pull request @vvvm23?
```
# load PT weights in FX
model = FlaxLlamaForCausalLM.from_pretrained("openlm-research/open_llama_3b_v2", from_pt=True)
# push FX weights to Hub
model.push_to_hub("openlm-research/open_llama_3b_v2", create_pr=True)
```
Once the PR is finalised by @ArthurZucker's review, we can merge this PR on the Hub and thus fix the failing doc tests! Thanks @vvvm23!

ayaka14732 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+        all_attentions = () if output_attentions else None
+        all_hidden_states = () if output_hidden_states else None
+
+        for block in self.blocks:
+            if output_hidden_states:
+                all_hidden_states += (hidden_states,)
+            layer_outputs = block(
+                hidden_states,
+                attention_mask=attention_mask,
+                position_ids=position_ids,
+                deterministic=deterministic,
+                init_cache=init_cache,
+                output_attentions=output_attentions,
+            )
+            hidden_states = layer_outputs[0]
+
+            if output_attentions:
+                all_attentions += (layer_outputs[1],)
```
`jax.lax.scan` should be used here. The for loop takes a very long time to compile.

vvvm23 replied:
It's a good idea, current approach was just to match the existing GPT Neo implementation. @sanchit-gandhi what do you think?

sanchit-gandhi replied:
Hmm both agree and disagree! For both inference and training, `jax.lax.scan` gives **significant** speed-ups to the compilation time. For training, it also gives a big memory and speed gain, since the compiler handles the N-layer transformer block very efficiently in the forward and backward pass. However for inference, `jax.lax.scan` is typically **slower** than a native `for i` loop (after you've done the compile step).

Given that Transformers mainly promotes JAX for inference, and `jax.lax.scan` adds a lot of code complexity, we've opted not to add `jax.lax.scan` into the main library. Here's a PR on how you could implement it for your own purposes though: #18341. You can see the huge complexity adding `jax.lax.scan` introduces to what is otherwise very clean modelling code. It's for this reason that we currently advise against adding it directly.

borisdayma replied:
In my experience scanning layers compiles much faster but can be slower and require more memory for both training and inference (sometimes it's barely slower for training but much slower for inference).
However you can use the `unroll` arg which makes it work like a `for` loop. The main inconvenient is that code can be a bit more complex with it.

ArthurZucker commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+            dtype=self.dtype,
+        )
+        self.layers = FlaxLlamaLayerCollection(self.config, dtype=self.dtype)
+        self.norm = FlaxLlamaRMSNorm(self.config)
```
should the dtype not be passed here as well?

ArthurZucker commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+        batch_size, seq_length = input_ids.shape
+
+        past_key_values = self.init_cache(batch_size, max_length)
+        # Note that usually one would have to put 0's in the attention_mask for x > input_ids.shape[-1] and x < cache_length.
```
nice 😉

ArthurZucker reviewed (approved):
Very clean PR thanks for this! 
Left a few nits here and there !

vvvm23 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+"""
+
+
+def create_sinusoidal_positions(num_pos, dim):
+    inv_freq = 1.0 / (10000 ** (np.arange(0, dim, 2) / dim))
+    freqs = np.einsum("i , j -> i j", np.arange(num_pos), inv_freq).astype("float32")
+
+    emb = np.concatenate((freqs, freqs), axis=-1)
+    out = np.concatenate((np.sin(emb)[:, None, :], np.cos(emb)[:, None, :]), axis=-1)
+    return jnp.array(out[:, :, :num_pos])
+
+
+def rotate_half(tensor):
+    """Rotates half the hidden dims of the input."""
+    rotate_half_tensor = jnp.concatenate(
+        (-tensor[..., tensor.shape[-1] // 2 :], tensor[..., : tensor.shape[-1] // 2]), axis=-1
+    )
+    rotate_half_tensor = jnp.concatenate(
+        (-tensor[..., tensor.shape[-1] // 2 :], tensor[..., : tensor.shape[-1] // 2]), axis=-1
+    )
+    return rotate_half_tensor
```
I don't think so - however I seem to do the same operation twice here for some reason :thinking: will remove one

vvvm23 replied:
@ArthurZucker I can't find functions with that name in either of those flax implementations. Going to resolve this for now.

vvvm23 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+            kernel_init=jax.nn.initializers.normal(self.config.initializer_range),
+        )
+
+        self.q_proj, self.k_proj, self.v_proj = dense(), dense(), dense()
```
I can't seem to find this flag in `configuration_llama.py` on this branch, could you point me to where it is?

vvvm23 replied:
Found it :)

vvvm23 commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,718 @@
+        )
+
+        hidden_states = outputs[0]
+        lm_logits = self.lm_head(hidden_states)
```
Did you mean to highlight these lines? I can't see any relation with dtype here.

sanchit-gandhi replied:
That's a great point @ArthurZucker - we should do the same here if LLaMA PyTorch does @vvvm23 to maintain 1-to-1 equivalence, especially for the largest checkpoints this will make a difference. You can see a code-snippet for BLOOM here: https://github.com/huggingface/transformers/blob/a5e6df82c00cb53ffe863008cbbedd813bcc508b/src/transformers/models/bloom/modeling_flax_bloom.py#L278

vvvm23 committed:
applies some suggestions from reviews

vvvm23 committed:
casts attn output logits to f32 regardless of dtype

vvvm23 commented on tests/models/llama/test_modeling_flax_llama.py:
```diff
@@ -0,0 +1,260 @@
+        inputs_dict = {"input_ids": input_ids, "attention_mask": attention_mask}
+        return config, inputs_dict
+
+    def check_use_cache_forward(self, model_class_name, config, input_ids, attention_mask):
```
probably :) I'm seeing a few functions around the codebase with the same code but no "Copied from" comment. Should this be added or fine without?

ArthurZucker replied:
if it's copied from let's add it

vvvm23 replied:
Done :+1:

vvvm23 replied:
should work now

vvvm23 committed:
adds attn bias using `LlamaConfig.attention_bias`

vvvm23 commented:
@sanchit-gandhi I forgot to say but the PR is up for the weights. Should be good to go once those PRs are merged.

vvvm23 committed:
adds Copied From comments to Flax Llama test

vvvm23 committed:
mistral and persimmon test change -copy from llama

vvvm23 commented:
@sanchit-gandhi @ArthurZucker a new issue has arisen. The tests fail when running `utils/check_docstrings.py` which prompts to run `make fix-copies`. However, this does not resolve the error. Any advice?

vvvm23 commented:
Hey @sanchit-gandhi @ArthurZucker just nudging on the above point, keen to wrap things up here!

To be more clear on my last, the error is:
```
Traceback (most recent call last):
  File "utils/check_docstrings.py", line 1239, in <module>
    check_docstrings(overwrite=args.fix_and_overwrite)
  File "utils/check_docstrings.py", line 1231, in check_docstrings
    raise ValueError(error_message)
ValueError: There was at least one problem when checking docstrings of public objects.
The following objects docstrings do not match their signature. Run `make fix-copies` to fix this.
- FlaxLlamaForCausalLM
- FlaxLlamaModel

Exited with code exit status 1
```

But running `make fix-copies` does not resolve the error.

vvvm23 committed:
updates docs index

vvvm23 committed:
removes Copied from in tests

it was preventing `make fix-copies` from succeeding

vvvm23 committed:
quality and style

ArthurZucker commented:
Hey! I was off for a week let me try to help you here! 😉

ArthurZucker commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,719 @@
+        )
+        return unfreeze(init_variables["cache"])
+
+    @add_start_docstrings_to_model_forward(LLAMA_INPUTS_DOCSTRING)
+    def __call__(
```
make fix copies might be failing on this one

ArthurZucker replied:
also params and some of the args here are not present in the `LLAMA_INPUTS_DOCSTRING`

vvvm23 replied:
interesting, I copied this from the Flax GPTNeo code. It also has these extra arguments that are not in the docstring. I will try with, but it is interesting this other example works without.

vvvm23 replied:
seems for GPTNeo, `utils/check_docstrings.py` ignores it. I will try the same with this model and see if it works.

ArthurZucker commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,719 @@
+        return model_kwargs
+
+
+append_call_sample_docstring(FlaxLlamaForCausalLM, _CHECKPOINT_FOR_DOC, FlaxCausalLMOutput, _CONFIG_FOR_DOC)
```
Or here

ArthurZucker reviewed (commented):
- the `test_pr_documentation_test` needs the correct path/acces to the model
- the `check_repository_consistency` needs additional changes to the docstring added to the `__call__` seems wrong and is not properly fixed. I'll check again if it's a bug on the coverage or not

vvvm23 committed:
ignores FlaxLlama input docstring

vvvm23 commented:
@ArthurZucker I added two lines that made the docstring checks ignore the Llama Flax models. This matches what GPTNeo did for their Flax models, so I assuming it is okay.

This (unsurprisingly) fixes that failing test. A few others have cropped up, but they seem unrelated to my changes. I will sync the branch with main and see if that fixes them. Other than that, and on the PRs in the individual model checkpoint hub pages, we should be good to go.

vvvm23 committed:
Merge branch 'main' into add-llama-flax

vvvm23 commented:
Merging with main fixed all the tests apart from the model checkpoint one. Should be good to go.

ArthurZucker commented:
For the last test you can use `from_pretrained("path", revision = "refs/pr/4")`

vvvm23 commented:
Hi @ArthurZucker , the test that fails is the pr documentation test. The checkpoint can only be specified via `CHECKPOINT_FOR_DOC` which doesn't allow for setting revisions. Any other ideas?

vvvm23 commented:
PRs on the hub would be better, however I cannot merge these PRs :sweat_smile: 

We could always create our own internal copy and point the test at that?

ArthurZucker commented:
I'll open a PR to support revisions anyway, another contributor is also stuck because of this

ArthurZucker commented:
Sorry for the delay, #27645 will help 😉

vvvm23 commented:
nice! left some comments on that PR

ArthurZucker commented:
Merged 😉

vvvm23 committed:
Merge branch 'main' into add-llama-flax

vvvm23 committed:
adds revision to `_CHECKPOINT_FOR_DOC`

vvvm23 committed:
repo consistency and quality

vvvm23 committed:
removes unused import

vvvm23 committed:
removes copied from from Phi test

now diverges from llama tests following FlaxLlama changes

vvvm23 commented:
The PR documentation test seems to time out. Is this a intermittent or known issue? Or something wrong with the code 🤔

vvvm23 commented:
@ArthurZucker @sanchit-gandhi could you advise on the above? Not intimately familiar with how your test workers operate.

ArthurZucker commented on src/transformers/models/llama/modeling_flax_llama.py:
```diff
@@ -0,0 +1,723 @@
+
+
+append_call_sample_docstring(
+    FlaxLlamaForCausalLM, _CHECKPOINT_FOR_DOC, FlaxCausalLMOutput, _CONFIG_FOR_DOC, revision="refs/pr/4"
```
the model is probably too big to be used, the checkpoint should be quantized!

ArthurZucker reviewed (commented):
Sorry for the delay, either use a smaller checkpoint, or just use a dummy checkpoint using something like this:
```python 
    @add_code_sample_docstrings(
        checkpoint=_CHECKPOINT_FOR_DOC,
        real_checkpoint=_REAL_CHECKPOINT_FOR_DOC,
        output_type=BaseModelOutputWithPast,
        config_class=_CONFIG_FOR_DOC,
    )
```
See here : 
https://github.com/huggingface/transformers/blob/5cd83cdbc6a55ce9b89697fc0d0818a7b6a52925/src/transformers/models/gpt_neox/modeling_gpt_neox.py#L45

vvvm23 committed:
Merge branch 'main' into add-llama-flax

vvvm23 committed:
adds `_REAL_CHECKPOINT_FOR_DOC`

kiansierra commented:
Hi @vvvm23 I believe that in order to pass the tests you should update the https://huggingface.co/afmck/testing-llama-tiny/tree/main to have the flax ckpts and  also the tokenizer, I created this https://huggingface.co/ksmcg/Mistral-tiny/tree/main for Mistral and it has passed the tests.
Also I belive the revision should be removed to pass

vvvm23 commented:
Ah good catch, I will fix that tomorrow

vvvm23 committed:
removes refs from pr tests

vvvm23 committed:
reformat to make ruff happy

vvvm23 commented:
Finally, green light 😁

ArthurZucker commented:
Let's just resolve the confilcts and good to go IMO!

vvvm23 committed:
Merge branch 'main' into add-llama-flax

vvvm23 commented:
done!

ArthurZucker commented:
Thanks for the PR 🤗
